// @tgwf/co2/dist/cjs/constants/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target, mod));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  DATACENTER_ENERGY: () => DATACENTER_ENERGY,
  END_USER_DEVICE_ENERGY: () => END_USER_DEVICE_ENERGY,
  FIRST_TIME_VIEWING_PERCENTAGE: () => FIRST_TIME_VIEWING_PERCENTAGE,
  GLOBAL_GRID_INTENSITY: () => GLOBAL_GRID_INTENSITY,
  KWH_PER_GB: () => KWH_PER_GB,
  NETWORK_ENERGY: () => NETWORK_ENERGY,
  PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD: () => PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD,
  PRODUCTION_ENERGY: () => PRODUCTION_ENERGY,
  RENEWABLES_GRID_INTENSITY: () => RENEWABLES_GRID_INTENSITY,
  RETURNING_VISITOR_PERCENTAGE: () => RETURNING_VISITOR_PERCENTAGE,
  fileSize: () => import_file_size.default
});
module.exports = __toCommonJS(constants_exports);
var import_file_size = __toESM(require("./file-size.js"));
var import_average_intensities_min = __toESM(require("../data/average-intensities.min.js"));
const KWH_PER_GB = 0.81;
const END_USER_DEVICE_ENERGY = 0.52;
const NETWORK_ENERGY = 0.14;
const DATACENTER_ENERGY = 0.15;
const PRODUCTION_ENERGY = 0.19;
const GLOBAL_GRID_INTENSITY = import_average_intensities_min.default.data["WORLD"];
const RENEWABLES_GRID_INTENSITY = 50;
const FIRST_TIME_VIEWING_PERCENTAGE = 0.75;
const RETURNING_VISITOR_PERCENTAGE = 0.25;
const PERCENTAGE_OF_DATA_LOADED_ON_SUBSEQUENT_LOAD = 0.02;
